events/ri/README.md

```diff
@@ -0,0 +1,137 @@
+
+### Topic partitioning and message ordering
+
+The `KafkaEventSubscriber` implementation illustrates how to partition the destination Kafka topic _by repository id and
+reference name_: this should allow downstream consumers to process events for a given reference in the order they were
+produced.
+
+Note that the order in which events are produced to a topic does not necessarily reflect the order in which Nessie's
+version store recorded them. As stated in Nessie's [Events API design document], strict ordering of events is not
+guaranteed. More specifically, for two events A and B that happened sequentially in a very short timeframe on the same
+reference, Nessie may notify the subscriber of B first, than A. In this case, the subscriber would put B in the topic
+before A, resulting in out-of-order sequencing.
+
+This only happens for very close events, and for most use cases this shouldn't be a problem; but if strict ordering is
+required, consider using more advanced techniques, such as [Kafka Streams] to re-order the messages, based e.g. on the
+commit timestamp.
+
+[Events API design document]:https://github.com/projectnessie/nessie/blob/nessie-0.60.1/design/events.md?plain=1#L364
+[Kafka Streams]:https://kafka.apache.org/documentation/streams/
+
+### Producer-side filtering
+
+Producer-side filtering is possible by changing the implementation of either the `getEventFilter()` or
+`getEventTypeFilter()` methods.
+
+To illustrate this, `KafkaEventSubscriber` uses its `EventTypeFilter` to completely ignore merge and transplant events.
+(Also note: merge and transplant events also trigger commit events, so it's still possible to capture what happened in a
+branch by listening to commit events only.)
+
+The subscriber is further configured to optionally watch just a few repository ids, or all of them. This is done by
+setting the `nessie.events.repository-ids` configuration option, and is implemented by modifying the subscriber's
+`EventFilter`. Again, this is purely illustrative. Other ways of filtering events on the subscriber side are possible,
+for example one could configure the subscribe to watch only certain branches or tags.
+
+### Using headers efficiently
+
+The `KafkaEventSubscriber` implementation also illustrates how to use Kafka headers to pass additional information along
+with the message. In this case, the subscriber adds the repository id, the user and the event creation timestamp to the
+message headers, allowing consumers to filter or sort messages based on these values.
+
+### Other considerations
+
+The `KafkaEventSubscriber` implementation also illustrates how to handle events in a non-blocking way, by publishing
+messages without waiting for the broker's acknowledgement. This is achieved by using the `KafkaProducer#send` method in
+a fire-and-forget mode, along with the `acks=0` producer configuration option. If your implementation needs to wait for
```

Comment:
As far as I know, using `acks=0` is not recommended in the majority of cases. I am not sure if using 0 in the reference implementation is the best solution.

Reply:
Sorry for the late reply. You are right, and I had to open my Kafka in Action book to refresh my skills around acks and send methods :-)

In short, we have 3 send methods:

1. fire-and-forget: don't wait for the future to complete, don't provide any callback;
2. async: don't wait for the future to complete, but provide a completion callback;
3. sync: wait for the future to complete (synchronously).

With Nessie, 1 and 2 are fine, but 3 should be avoided (or at least, `isBlocking()` should return true).

The way these methods really work will further depend on the `acks` setting (and `retries`). In short, the returned future is completed when:

1. With acks=0 => immediately when the message is sent;
2. With acks=1 => when the broker acknowledged the write;
3. With acks=all => when the broker and all the in-sync replicas acknowledged the write.

Which brings us to the conclusion: the number of acks is relatively orthogonal to the send method. 

With all that in mind, I will indeed change the example to use the async send method AND ack=1, since both seem reasonable choices for this use case.